rs/messaging/tests/memory_tests.rs

```diff
@@ -146,15 +162,102 @@ fn check_guaranteed_response_message_memory_limits_are_respected_impl(
     // One extra tick to make sure everything is gc'ed.
     fixture.tick();
 
-    // Check the system agrees on 'no hanging calls'.
-    if fixture.open_call_contexts_count().values().sum::<usize>() != 0 {
-        return fixture.failed_with_reason("found call contexts after shutdown phase");
+    // Check the records agree on 'no hanging calls'.
+    if fixture
+        .canisters()
+        .into_iter()
+        .map(|canister| extract_metrics(&fixture.force_query_records(canister)))
+        .any(|metrics| metrics.hanging_calls != 0)
+    {
+        return fixture.failed_with_reason("found hanging calls in the records");
     }
 
     // After the fact, all memory is freed and back to 0.
     fixture.expect_guaranteed_response_message_memory_taken_at_most("Final check", 0, 0)
 }
 
+/// Runs a state machine test with two subnets, a local subnet with 2 canisters installed and a
+/// remote subnet with 5 canisters installed. All canisters, except one local canister referred to
+/// as `migrating_canister` is stopped.
+///
+/// In the first phase a number of rounds are executed on both subnets, including XNet traffic with
+/// the `migrating_canister` making random calls to all installed canisters (since all calls are
+/// rejected except those to self, downstream calls are disabled).
+///
+/// For the second phase, `migrating_canister` stops making calls and is then migrated to the
+/// remote subnet. Since all other canisters are stopped, there are bound to be a number of reject
+/// signals for requests in the reverse stream to the local_subnet. But since we migrated the
+/// `migrating_canister` to the remote subnet, the locally generated reject responses fail to
+/// induct and are rerouted into the stream to the remote subnet. The remote subnet eventually picks
+/// them up and inducts them into `migrating_canister` leaving no hanging calls after some more rounds.
+///
+/// If there are hanging calls after a threshold number of rounds, there is most likely a bug
+/// connected to reject signals for requests, specifically with the corresponding exceptions due to
+/// canister migration.
+#[test]
+fn check_calls_conclude_with_migrating_canister() {
```

Comment:
One general comment on the files in 'rs/messaging/tests/'. The names are arbitrary and perhaps this should be refactored into just one file 'rs/messaging/tests/messaging.rs' since this test isn't a 'memory test', it's also not exactly clear what a 'queue test' is supposed to be and why is it in message routing when queues are in replicated state and so on.

What I made here is such that what I want to check (the rerouting of locally generated reject responses into the stream to the subnet where the canister moved to) but also such that I can just use the framework I already have. In principle there is nothing wrong with it.

But from what we were discussing and what might also be interesting to see in action is a great deal of such reject responses ending up in one stream by making one canister send messages to lots canister on lots of subnets, and then checking whether all of the corresponding reject responses land in the same stream. I mean this test does check that but with just one subnet. I didn't go for this because I might have to extend the framework quite a bit and also it would make the test run for a long time just to generate so many messages so I figured it's not worth it.